pkg/workspace/resource/node_claim.go

```diff
@@ -0,0 +1,329 @@
+	return &NodeClaimManager{
+		Client:       c,
+		recorder:     recorder,
+		expectations: expectations,
+		logger:       klog.NewKlogr().WithName("NodeClaim"),
+	}
+}
+
+func (c *NodeClaimManager) SyncNodeClaims(ctx context.Context, wObj *kaitov1beta1.Workspace) (bool, error) {
+	if err := c.ensureNodeClaims(ctx, wObj); err != nil {
+		return false, err
+	}
+
+	return c.areNodeClaimsReady(ctx, wObj)
+}
+
+// ensureNodeClaims ensures the correct number of NodeClaims for the workspace
+// based on the TargetNodeCount in the workspace status, considering preferred nodes
+func (c *NodeClaimManager) ensureNodeClaims(ctx context.Context, wObj *kaitov1beta1.Workspace) error {
+	workspaceKey := client.ObjectKeyFromObject(wObj).String()
+
+	if !c.expectations.SatisfiedExpectations(c.logger, workspaceKey) {
+		klog.V(4).InfoS("Waiting for NodeClaim expectations to be satisfied",
+			"workspace", klog.KObj(wObj))
+		return nil
+	}
+
+	// Calculate the number of NodeClaims needed (target - preferred nodes)
+	requiredNodeClaimsCount, err := nodeclaim.GetRequiredNodeClaimsCount(ctx, c.Client, wObj)
+	if err != nil {
+		return fmt.Errorf("failed to get required NodeClaims: %w", err)
+	}
+
+	klog.V(4).InfoS("NodeClaim calculation",
+		"workspace", klog.KObj(wObj),
+		"requiredNodeClaims", requiredNodeClaimsCount)
+
+	existingNodeClaims, err := nodeclaim.GetExistingNodeClaims(ctx, c.Client, wObj)
+	if err != nil {
+		return fmt.Errorf("failed to get existing NodeClaims: %w", err)
+	}
+
+	currentNodeClaimCount := len(existingNodeClaims)
+
+	if currentNodeClaimCount < requiredNodeClaimsCount {
+		nodesToCreate := requiredNodeClaimsCount - currentNodeClaimCount
+		klog.InfoS("Creating additional NodeClaims",
+			"workspace", klog.KObj(wObj),
+			"current", currentNodeClaimCount,
+			"required", requiredNodeClaimsCount,
+			"toCreate", nodesToCreate)
+
+		if updateErr := workspace.UpdateStatusConditionIfNotMatch(ctx, c.Client, wObj, kaitov1beta1.ConditionTypeNodeClaimStatus, metav1.ConditionFalse,
+			"CreatingNodeClaims", fmt.Sprintf("Creating %d additional NodeClaims (current: %d, required: %d)", nodesToCreate, currentNodeClaimCount, requiredNodeClaimsCount)); updateErr != nil {
+			klog.ErrorS(updateErr, "failed to update NodeClaim status condition", "workspace", klog.KObj(wObj))
+		}
+
+		c.expectations.ExpectCreations(c.logger, workspaceKey, nodesToCreate)
+
+		nodeOSDiskSize := c.determineNodeOSDiskSize(wObj)
+
+		for range nodesToCreate {
+			var nodeClaim *karpenterv1.NodeClaim
+			var err error
+			created := false
+
+			retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				nodeClaim = nodeclaim.GenerateNodeClaimManifest(nodeOSDiskSize, wObj)
+				err = c.Client.Create(ctx, nodeClaim)
+
+				if err == nil {
+					created = true
+					return nil
+				}
+				return err
+			})
+
+			if !created {
+				// Failed to create, decrement expectations
+				c.expectations.CreationObserved(c.logger, workspaceKey)
```

Comment:
I think you need to add expectation one by one. 

now you have `c.expectations.ExpectCreations(c.logger, workspaceKey, nodesToCreate)`
if the first nodeclaim failed to created,  you will only decrease once.

Reply:
fixed, create/delete nodeclaim will continue when the create/delete request fails.